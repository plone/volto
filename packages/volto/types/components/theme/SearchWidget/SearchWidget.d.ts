export default SearchWidget;
declare function SearchWidget(props: any): import("react/jsx-runtime").JSX.Element;
