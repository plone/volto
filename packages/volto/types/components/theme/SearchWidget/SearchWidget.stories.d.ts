export const SearchWidget: any;
declare namespace _default {
    export let title: string;
    export { SearchWidget as component };
    export let decorators: ((Story: any) => JSX.Element)[];
}
export default _default;
