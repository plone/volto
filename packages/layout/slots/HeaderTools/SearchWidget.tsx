import { type FormEvent, useState } from 'react';
import { Form, Input, Label, TextField } from 'react-aria-components';
import type { SlotComponentProps } from '../SlotRenderer';
import { useNavigate } from 'react-router';
import { useTranslation } from 'react-i18next';
import { Icon } from '@plone/components';
import { Button } from '@plone/components/quanta';
import styles from './Tools.module.css';
import clsx from 'clsx';

const SearchWidget = (props: SlotComponentProps) => {
  const { location } = props;
  const [searchQuery, setSearchQuery] = useState('');

  const { t } = useTranslation();
  const navigate = useNavigate();
  const pathname = location.pathname;

  const onSubmit = (e: FormEvent<HTMLFormElement>) => {
    const path =
      pathname?.length > 0 ? `&path=${encodeURIComponent(pathname)}` : '';

    navigate(
      `/search?SearchableText=${encodeURIComponent(searchQuery)}${path}`,
    );

    setSearchQuery('');
    e.preventDefault();
  };

  return (
    <Form
      className={clsx(styles['searchwidget'], 'searchwidget')}
      onSubmit={onSubmit}
    >
      <TextField name="search" type="text">
        <Label className="sr-only">
          {t('layout.slots.tools.searchWidget.searchSite')}
        </Label>
        <Input
          placeholder={t('layout.slots.tools.searchWidget.searchSite')}
          onChange={(e) => setSearchQuery(e.target.value)}
          value={searchQuery}
        />
      </TextField>
      <Button type="submit">
        <Icon size="sm">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 36 36">
            <path
              fillRule="evenodd"
              d="M7,16 C7,11.038 11.037,7 16,7 C20.963,7 25,11.038 25,16 C25,20.962 20.963,25 16,25 C11.037,25 7,20.962 7,16 L7,16 Z M32.707,31.293 L24.448,23.034 C26.039,21.125 27,18.673 27,16 C27,9.935 22.065,5 16,5 C9.935,5 5,9.935 5,16 C5,22.065 9.935,27 16,27 C18.673,27 21.125,26.039 23.034,24.448 L31.293,32.707 L32.707,31.293 Z"
            />
          </svg>
        </Icon>
      </Button>
    </Form>
  );
};

export default SearchWidget;
