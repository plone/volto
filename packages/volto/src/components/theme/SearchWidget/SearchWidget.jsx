import { useState, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { Form, Input } from 'semantic-ui-react';
import { defineMessages, useIntl } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';

import { getNavroot } from '@plone/volto/actions';
import { hasApiExpander, getBaseUrl } from '@plone/volto/helpers';
import { Icon } from '@plone/volto/components';
import zoomSVG from '@plone/volto/icons/zoom.svg';

const messages = defineMessages({
  search: {
    id: 'Search',
    defaultMessage: 'Search',
  },
  searchSite: {
    id: 'Search Site',
    defaultMessage: 'Search Site',
  },
});

const SearchWidget = (props) => {
  const intl = useIntl();
  const dispatch = useDispatch();

  const [text, setText] = useState('');
  const history = useHistory();
  const onChangeText = (event, { value }) => {
    setText(value);
  };
  const pathname = props.pathname;
  const onSubmit = (event) => {
    const path =
      pathname?.length > 0 ? `&path=${encodeURIComponent(pathname)}` : '';

    history.push(`./search?SearchableText=${encodeURIComponent(text)}${path}`);
    // reset input value
    setText('');
    event.preventDefault();
  };

  const navroot = useSelector((state) => state.navroot?.data);
  useEffect(() => {
    if (!hasApiExpander('navroot', getBaseUrl(pathname))) {
      dispatch(getNavroot(getBaseUrl(pathname)));
    }
  }, [dispatch, pathname]);

  return (
    <Form action={`${navroot?.navroot?.['@id']}/search`} onSubmit={onSubmit}>
      <Form.Field className="searchbox">
        <Input
          aria-label={intl.formatMessage(messages.search)}
          onChange={onChangeText}
          name="SearchableText"
          value={text}
          transparent
          autoComplete="off"
          placeholder={intl.formatMessage(messages.searchSite)}
          title={intl.formatMessage(messages.search)}
        />
        <button aria-label={intl.formatMessage(messages.search)}>
          <Icon name={zoomSVG} size="18px" />
        </button>
      </Form.Field>
    </Form>
  );
};

export default SearchWidget;
