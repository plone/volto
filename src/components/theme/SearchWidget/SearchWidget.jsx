/**
 * Search widget component.
 * @module components/theme/SearchWidget/SearchWidget
 */

import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { Form, Input } from 'semantic-ui-react';
import { compose } from 'redux';
import { PropTypes } from 'prop-types';
import { defineMessages, injectIntl } from 'react-intl';

import { Icon } from '@plone/volto/components';
import zoomSVG from '@plone/volto/icons/zoom.svg';

const messages = defineMessages({
  search: {
    id: 'Search',
    defaultMessage: 'Search',
  },
  searchSite: {
    id: 'Search Site',
    defaultMessage: 'Search Site',
  },
});

/**
 * SearchWidget component class.
 * @class SearchWidget
 * @extends Component
 */
class SearchWidget extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    pathname: PropTypes.string,
  };

  /**
   * Constructor
   * @method constructor
   * @param {Object} props Component properties
   * @constructs WysiwygEditor
   */
  constructor(props) {
    super(props);
    this.onChangeText = this.onChangeText.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.state = {
      text: '',
    };
  }

  /**
   * On change text
   * @method onChangeText
   * @param {object} event Event object.
   * @param {string} value Text value.
   * @returns {undefined}
   */
  onChangeText(event, { value }) {
    this.setState({
      text: value,
    });
  }

  /**
   * Submit handler
   * @method onSubmit
   * @param {event} event Event object.
   * @returns {undefined}
   */
  onSubmit(event) {
    const path =
      this.props.pathname?.length > 0
        ? `&path=${encodeURIComponent(this.props.pathname)}`
        : '';
    this.props.history.push(
      `/search?SearchableText=${encodeURIComponent(this.state.text)}${path}`,
    );
    event.preventDefault();
  }

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    return (
      <Form action="/search" onSubmit={this.onSubmit}>
        <Form.Field className="searchbox">
          <Input
            aria-label={this.props.intl.formatMessage(messages.search)}
            onChange={this.onChangeText}
            name="SearchableText"
            value={this.state.text}
            transparent
            autoComplete="off"
            placeholder={this.props.intl.formatMessage(messages.searchSite)}
            title={this.props.intl.formatMessage(messages.search)}
          />
          <button aria-label={this.props.intl.formatMessage(messages.search)}>
            <Icon name={zoomSVG} size="18px" />
          </button>
        </Form.Field>
      </Form>
    );
  }
}

export default compose(withRouter, injectIntl)(SearchWidget);
