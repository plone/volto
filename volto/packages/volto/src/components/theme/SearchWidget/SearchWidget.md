Search example, with `intl` auto-injected:

```jsx static
<SearchWidget pathname="" />
```

Output:

```jsx noeditor
const { Provider } = require('react-intl-redux');
const configureStore = require('redux-mock-store').default;
const store = configureStore()({
  userSession: {
    login: {},
  },
  intl: {
    locale: 'en',
    messages: {},
  },
});

<div className={'rsg--pre-42'}>
  <Provider store={store}>
    <SearchWidget pathname="" />
  </Provider>
</div>;
```
