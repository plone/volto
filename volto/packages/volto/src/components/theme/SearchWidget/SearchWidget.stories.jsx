import { injectIntl } from 'react-intl';
import React from 'react';
import SearchWidgetComponent from './SearchWidget';
import { RealStoreWrapper as Wrapper } from '@plone/volto/storybook';

const IntlSearchWidgetComponent = injectIntl(SearchWidgetComponent);

function StoryComponent(args) {
  return (
    <Wrapper
      customStore={{
        intl: {
          locale: 'en',
          messages: {},
        },
      }}
    >
      <div id="toolbar" style={{ display: 'none' }} />
      <IntlSearchWidgetComponent {...args} pathname="/blog" />
    </Wrapper>
  );
}

export const SearchWidget = StoryComponent.bind({});

export default {
  title: 'Public components/SearchWidget',
  component: SearchWidget,
  decorators: [
    (Story) => (
      <div className="ui segment form attached" style={{ width: '400px' }}>
        <Story />
      </div>
    ),
  ],
};
