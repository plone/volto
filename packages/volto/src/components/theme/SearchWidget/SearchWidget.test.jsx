import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-intl-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';

import SearchWidget from './SearchWidget';

const mockStore = configureStore();

describe('SearchWidget', () => {
  it('renders a search widget component', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
      navroot: {
        data: {
          '@id': 'http://localhost:3000/@navroot',
          navroot: {
            '@id': 'http://localhost:3000',
          },
        },
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <SearchWidget pathname="/blog" />
        </MemoryRouter>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
