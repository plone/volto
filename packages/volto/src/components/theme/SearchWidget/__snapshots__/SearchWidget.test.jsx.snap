// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchWidget renders a search widget component 1`] = `
<form
  action="http://localhost:3000/search"
  className="ui form"
  onSubmit={[Function]}
>
  <div
    className="field searchbox"
  >
    <div
      className="ui transparent input"
    >
      <input
        aria-label="Search"
        autoComplete="off"
        name="SearchableText"
        onChange={[Function]}
        placeholder="Search Site"
        title="Search"
        type="text"
        value=""
      />
    </div>
    <button
      aria-label="Search"
    >
      <svg
        className="icon"
        dangerouslySetInnerHTML={
          Object {
            "__html": undefined,
          }
        }
        onClick={null}
        style={
          Object {
            "fill": "currentColor",
            "height": "18px",
            "width": "auto",
          }
        }
        viewBox=""
        xmlns=""
      />
    </button>
  </div>
</form>
`;
